rllib/utils/metrics/stats/base.py

```diff
@@ -0,0 +1,291 @@
+            is_leaf = state["is_leaf"]
+            return {
+                "is_root": is_root,
+                "is_leaf": is_leaf,
+            }
+        elif stats_object is not None:
+            return {
+                "is_root": stats_object.is_root,
+                "is_leaf": stats_object.is_leaf,
+            }
+        else:
+            raise ValueError("Either stats_object or state must be provided")
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        ...
+
+    @abstractmethod
+    def merge(self, incoming_stats: List["StatsBase"]) -> None:
+        """Merges StatsBase objects.
+
+        Args:
+            incoming_stats: The list of StatsBase objects to merge.
+        """
+        ...
+
+    @abstractmethod
+    def push(self, value: Any) -> None:
+        """Pushes a value into this Stats object.
+
+        Args:
+            value: The value to push. Can be of any type.
+                GPU tensors are moved to CPU memory.
+        """
+        assert (
+            self.is_leaf
+        ), "Cannot push values to non-leaf Stats. Non-leaf Stats can only receive values via merge()."
+        ...
+
+    @abstractmethod
+    def peek(
+        self, compile: bool = True, latest_merged_only: bool = False
+    ) -> Union[Any, List[Any]]:
+        """Returns the result of reducing the internal values list.
+
+        Note that this method does NOT alter the internal values list in this process.
+        Thus, users can call this method to get an accurate look at the reduced value(s)
+        given the current internal values list.
+
+        Args:
+            compile: If True, the result is compiled into a single value if possible.
+            latest_merged_only: If True, only considers the latest merged values.
+                This parameter only works on aggregation stats objects (is_leaf=False).
+                When enabled, peek() will only use the values from the most recent merge operation.
+
+        Returns:
+            The result of reducing the internal values list on CPU memory.
+        """
+
+    @abstractmethod
+    def reduce(self, compile: bool = True) -> Union[Any, "StatsBase"]:
+        """Reduces the internal values.
+
+        This method should NOT be called directly by users.
+        It can be used as a hook to prepare the stats object for sending it to the root metrics logger and starting a new 'reduce cycle'.
+
+        The reduction logic depends on the implementation of the subclass.
+        Meaning that some classes may reduce to a single value, while others do not or don't even contain values.
+
+        Args:
+            compile: If True, the result is compiled into a single value if possible.
+                If False, the result is a Stats object similar to itself, but with the internal values reduced.
+        Returns:
+            The reduced value or a Stats object similar to itself, but with the internal values reduced.
+        """
+        ...
```

Comment:
With docstrings, you technically don't need `...` for abstract methods like `peek` above

Reply:
Thanks!